roles/openshift_health_checker/openshift_checks/docker_image_availability.py

```diff
@@ -0,0 +1,148 @@
+        if not registries:
+            return {"failed": True, "msg": "Unable to read docker registry data from host."}
+
+        # inspect all rpm or containerized docker images for every docker registry found
+        while registries and required_images:
+            registry = registries.pop(0)
+            required_images = self.inspect_images(registry, required_images, task_vars)
+
+        # if there are required images left, fail the check
+        if required_images:
+            msg = "Unable to find required images in any docker registries: {imgs}".format(imgs=list(required_images))
+            return {"failed": True, "changed": True, "msg": msg}
+
+        return {"changed": True}
+
+    def get_required_images(self, is_containerized, reg_prefix, openshift_release):
+        if is_containerized:
+            return set(self.containerized_docker_images(reg_prefix, openshift_release))
+
+        return set(self.rpm_docker_images(reg_prefix, openshift_release))
+
+    def containerized_docker_images(self, reg_prefix, version):
+        return [
+            "{image}:{version}".format(image=self.image_from_reg_prefix(reg_prefix), version=version)
+        ]
+
+    @staticmethod
+    def rpm_docker_images(reg_prefix, version):
+        return [
+            "{image_repo}/registry-console:{version}".format(image_repo=reg_prefix["repo"], version=version)
+        ]
+
+    @staticmethod
+    def qualified_docker_images(image_name, version):
+        return [
+            "{image_name}-{component}:{version}".format(**locals())
+            for component in "haproxy-router docker-registry deployer pod".split()
+        ]
+
+    @staticmethod
+    def image_from_reg_prefix(prefix):
+        return "".join([prefix["repo"], "/", prefix["image"]])
+
+    def inspect_images(self, registry, tagged_images, task_vars):
+        failed = set()
+        for image in tagged_images:
+            # the run command pulls the {skopeo_image} if it has not already been pulled.
+            # if pulling this image fails, the current image will be added to the failed set.
+            cmd_str = "docker run {skopeo_image} skopeo inspect docker://{registry}/{tagged_image}".format(
+                skopeo_image=self.skopeo_image,
+                registry=registry,
+                tagged_image=image
+            )
+
+            # use a "docker run ..." command on the remote host to inspect each specific
+            # image inside of a Skopeo image container. If the command's return code is
+            # non-zero, it is safe to assume that an image with the specified tag does not
+            # exist in the given registry.
+            if self.exec_cmd(cmd_str, task_vars)["rc"] != 0:
+                failed.add(image)
+
+        return failed
+
+    def inspect_local_images(self, tagged_images, task_vars):
+        failed = set()
+        for image in tagged_images:
+            cmd = self.module_executor("docker_image_facts", {"name": image}, task_vars)
```

Comment:
The pastebin is empty now (and on the intranet...), I missed it.

See https://github.com/openshift/openshift-ansible/pull/3461/files#r106359296 (a proposed sketch for the `run` method) and it might become easier to implement ;-)

```python
def local_images(self, names, task_vars):
    result = self.module_executor("docker_image_facts", {"name": names}, task_vars)
    if result.get("failed", False):
        raise OpenShiftCheckException(...)
    return [... for image in result["images"]]
```

The documented return values are not very clear to me http://docs.ansible.com/ansible/docker_image_facts_module.html#return-values.
We'll have to be careful here and do some tests:

1. If `docker_image_facts` work like `docker inspect`, then a name like "openshift/origin" may get translated into "docker.io/openshit/origin:latest".
2. `docker inspect` (command line) returns an empty list and an error if one of the images is not available:

```
$ docker inspect foobar busybox
[]
Error: No such image, container or task: foobar
```

More than that, it actually fails on the first error...

```
$ docker inspect busybox foobar | tail
Error: No such image, container or task: foobar
            }
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:38ac8d0f5bb30c8b742ad97a328b77870afaec92b33faf7e121161bc78a3fec8"
            ]
        }
    }
]
```

Perhaps you already knew all that and thus was proposing to get facts one by one?